### ISSUE TITLE
Run all SPARQL spec tests

### ISSUE THREAD

**Issue by rubensworks** [comunica/comunica#721]
#### Issue type:

- :heavy_plus_sign: Feature request

____
#### Description:

The package.json file of engines/query-sparql contains several `spec:*` scripts to invoke [SPARQL spec tests](https://github.com/w3c/rdf-tests/blob/gh-pages/sparql11/data-sparql11/manifest-all.ttl).
We should make sure that all of them are invoked in our GitHub CI workflow:

* [x] `spec:query`: http://www.w3.org/TR/sparql11-query/
* [x] `spec:update`: http://www.w3.org/TR/sparql11-update/ (depends on #435)
* [ ] `spec:csv-tsv`: http://www.w3.org/TR/sparql11-results-csv-tsv/ (requires CSV/TSV parsers)
* [ ] `spec:json`: http://www.w3.org/TR/sparql11-results-json/ (should be enableable directly)
* [ ] `spec:fed`: http://www.w3.org/TR/sparql11-federated-query/ (requires support in rdf-test-suite)
* [ ] `spec:sd`: http://www.w3.org/TR/sparql11-service-description/ (?)
* [ ] `spec:prot`: http://www.w3.org/TR/sparql11-protocol/ (depends on #435)
* [ ] `spec:graphstore`: http://www.w3.org/TR/sparql11-http-rdf-update/ (depends on #435)

Each one that succeeds should be added to the `spec` script.

Some of these may require changes to https://github.com/rubensworks/rdf-test-suite.js

Afterwards, we should share our earl report on https://github.com/w3c/rdf-tests/

It's possible to take up this bounty in smaller parts.

-----

## Bounty

A [bounty](https://comunica.dev/association/bounties/) has been placed on this issue by:

| |
|-|
| [<img alt="Comunica Association" src="https://comunica.dev/img/comunica_red.svg" height="50">](https://comunica.dev/association/) |
| *Variable budget* |

[Click here to learn more](https://comunica.dev/association/bounty_process/) if you're interested in claiming this bounty by resolving this issue.

**Comment by noahvsb** [comment]
The SPARQL spec test link is outdated. I think this is the updated one: https://github.com/w3c/rdf-tests/blob/main/sparql/sparql11/manifest-all.ttl

**Comment by kavixo** [comment]
Hi, I’m interested in taking this on.

**Comment by yui-stingray** [comment]
Hi, I would be interested in taking a small, well-scoped slice of this bounty if the association/company is open to assigning partial work.

Proposed slice: validate `spec:json-1-1` in `engines/query-sparql`, and if it passes cleanly, add it to the aggregate `spec` script with the corresponding validation evidence. If it does not pass cleanly, I would stop at a concise blocker report instead of broadening the scope.

Estimated time after assignment: within 1 day for the `spec:json-1-1` slice.
Budget: I am open to the association/company's expected budget for this partial slice.

I will not start implementation until assignment is confirmed, in line with the bounty procedure.

**Comment by voidten** [comment]
@rubensworks Looks like the AI swarm found this issue 😄. I have some time this week and would be happy to look at this.

**Comment by rubensworks** [comment]
@voidten Sure, but before we start, can you open an invoice for your completed bounties on OpenCollective? Since it's quite a substantial amount already.

And for this issue here, I would split the effort up into the different steps mentioned above.
Before you start with each step, can you provide an estimate of the time required before starting each step, with a short description of what would be done exactly?
To make sure we're aligned on what needs to be done, I'll then have a look at it before you actually get started.

**Comment by gamechangsu** [comment]
Hi @rubensworks, I scoped a small, reviewable slice of this bounty and would like to be considered if it is useful.

Proposed slice: enable the already-defined `spec:json-1-1` check in the aggregate `spec` script, with verification notes.

Scoping evidence from current `master`:
- `corepack yarn install --frozen-lockfile --ignore-scripts`
- `corepack yarn run build`
- `corepack yarn run build:engines`
- `corepack yarn --cwd engines/query-sparql run spec:json-1-1`

Result: `spec:json-1-1` passed with `4 / 4 tests succeeded`.

I also checked `spec:csv-tsv-1-1` while scoping. It exits successfully, but the TSV result cases are skipped because `text/tab-separated-values` expected results cannot be parsed, so I would not include CSV/TSV in this same small slice without first agreeing on parser work.

Estimated time after assignment: same day, roughly 1-2 hours for the focused PR and final verification.
Budget: open to the association/company's expected budget for this small partial slice.

I will not start implementation until assignment/confirmation, in line with the bounty procedure.

**Comment by voidten** [comment]
Estimate: 5 minutes.

`spec:json-1-1` already passes cleanly with `4 / 4` tests, so this step only adds it to the main `spec` script.

**Comment by rubensworks** [comment]
@voidten Since the json one is quite trivial as you mentioned, let's combine it with another one (up to you to choose which one).

**Comment by voidten** [comment]
Estimated time: about 1-2 hours.

I will combine it with `spec:csv-tsv-1-1` by first adding TSV parsing and CSV result-format support in `rdf-test-suite`.

**Comment by rubensworks** [comment]
> I will combine it with spec:csv-tsv-1-1 by first adding TSV parsing and CSV result-format support in rdf-test-suite.

@voidten Sounds good!
Concretely, we'd need rdf-test-suite to send the query to the engine and specifically request for TSV or CSV results.
The actual TSV/CSV comparison should then be done in rdf-test-suite. (this is because SPARQL/CSV is lossy and doesn't support round-tripping)

Let's go for a bounty of 320 euros for the `spec:csv-tsv-1-1` and `spec:json` parts together?

**Comment by voidten** [comment]
Sounds good 👍

**Comment by voidten** [comment]
@rubensworks I’ll work on `spec:fed` now. The federated query tests use `qt:serviceData` in the manifest, which `rdf-test-suite` doesn't support yet, so we'll need to address that first.

**Comment by rubensworks** [comment]
@voidten Sounds good! Can you share your estimated time before you start, so we can discuss the budget based on that?

**Comment by voidten** [comment]
The rdf-test-suite part will probably take me 1–1.5 hours. However, I suspect we’ll also need to make changes to Comunica (similar to the CSV/TSV work), and I can’t tell what those will be until the first part is done.

**Comment by rubensworks** [comment]
Ok, sounds good. Let's also reserve a budget of 320 euros for `spec:fed` then?

If there are straightforward fixes to be done in Comunica wrt federation, this can be part of this bounty indeed.
But if there are major problems in the implementation, then we can keep this out of scope.
Especially the queries with variable `SERVICE` targets won't be supported yet, so we should not implement them here (see https://github.com/comunica/comunica/issues/1384 )

**Comment by voidten** [comment]
> `spec:sd`: http://www.w3.org/TR/sparql11-service-description/ (the tests don't do anything?)

@rubensworks The tests are essentially conformance requirements for an endpoint. We'd need to implement a `TestCaseServiceDescriptionHandler` in `rdf-test-suite` and implement those three checks there.

**Comment by rubensworks** [comment]
@voidten So you would just hardcode logic according to the `mf:name`? I guess we can do that, for the lack of a better solution.

**Comment by voidten** [comment]
Yes, exactly.

**Comment by voidten** [comment]
@rubensworks I’ll work on `spec:sd` now. It'll probably take about 2 hours.

**Comment by rubensworks** [comment]
@voidten Sounds good. Let's fix 160 euros for `spec:sd` then.

**Comment by voidten** [comment]
@rubensworks I’ll work on `spec:prot` now. This will require changes to `rdf-test-suite`, and I’m pretty sure there will be some changes needed in Comunica too. Estimated time: about 2–2.5 hours.

**Comment by rubensworks** [comment]
> @rubensworks I’ll work on spec:prot now. This will require changes to rdf-test-suite, and I’m pretty sure there will be some changes needed in Comunica too. Estimated time: about 2–2.5 hours.

@voidten Ok, great. Let's do 200 euros for that part.

**Comment by rubensworks** [comment]
@voidten If you're up for it, feel free to give an estimate of how long the graph store one would take you. This would have to include #1717. I don't know the spec that well myself, but it doesn't seem too much work. My only worry is that it could clash with our current SPARQL endpoint in some ways, so we'd need to plan this out carefully.
